src/EventTeachingService.cpp

```diff
@@ -11,259 +11,427 @@
+          {
+            module_config->cleareventEEPROM(e);
+          }
 
-      controller->sendWRACK();
-    }
+          // recreate the hash table
+          module_config->clearEvHashTable();
+          // DEBUG_SERIAL << F("> cleared all events") << endl;
 
-    return PROCESSED;
+          controller->sendWRACK();
+          controller->sendGRSP(OPC_NNCLR, getServiceID(), GRSP_OK);
+        }
+        else
+        {
+          controller->sendCMDERR(CMDERR_NOT_LRN);
+          controller->sendGRSP(OPC_NNCLR, getServiceID(), CMDERR_NOT_LRN);
+        }
+      }
 
-  case OPC_NNEVN:
-    // request for number of free event slots
+      return PROCESSED;
 
-    if (module_config->nodeNum == nn)
-    {
-      byte free_slots = 0;
+    case OPC_NNEVN:
+      // 56 - request for number of free event slots
 
-      // count free slots using the event hash table
-      for (byte i = 0; i < module_config->EE_MAX_EVENTS; i++)
+      if (module_config->nodeNum == nn) 
       {
-        if (module_config->getEvTableEntry(i) == 0)
+        byte free_slots = 0;
+
+        // count free slots using the event hash table
+        for (byte i = 0; i < module_config->EE_MAX_EVENTS; i++) 
         {
-          ++free_slots;
+          if (module_config->getEvTableEntry(i) == 0) 
+          {
+            ++free_slots;
+          }
         }
-      }
 
-      // DEBUG_SERIAL << F("> responding to to NNEVN with EVNLF, free event table slots = ") << free_slots << endl;
-      controller->sendMessageWithNN(OPC_EVNLF, free_slots);
-    }
+        // DEBUG_SERIAL << F("> responding to to NNEVN with EVNLF, free event table slots = ") << free_slots << endl;
+        controller->sendMessageWithNN(OPC_EVNLF, free_slots);
+      }
 
-    return PROCESSED;
+      return PROCESSED;
 
-  case OPC_EVLRN:
-  {
-    // received EVLRN -- learn an event
-    byte evindex = msg->data[5];
-    byte evval = msg->data[6];
+    case OPC_EVLRN:
+      // D2 - learn an event
 
-    // DEBUG_SERIAL << endl << F("> EVLRN for source nn = ") << nn << F(", en = ") << en << F(", evindex = ") << evindex << F(", evval = ") << evval << endl;
 
-    // we must be in learn mode
-    if (bLearn)
-    {
-      // search for this NN, EN as we may just be adding an EV to an existing learned event
-      // DEBUG_SERIAL << F("> searching for existing event to update") << endl;
-      byte index = module_config->findExistingEvent(nn, en);
+      // DEBUG_SERIAL << endl << F("> EVLRN for source nn = ") << nn << endl;
 
-      // not found - it's a new event
-      if (index >= module_config->EE_MAX_EVENTS)
+      // we must be in learn mode
+      if (bLearn) 
       {
-        // DEBUG_SERIAL << F("> existing event not found - creating a new one if space available") << endl;
-        index = module_config->findEventSpace();
+        if (msg->len < 7) 
+        {
+          controller->sendGRSP(OPC_EVLRN, getServiceID(), CMDERR_INV_CMD);
```

Comment:
Should we send a CMDERR for short messages?

Reply:
Again, this is just the opcode and node number, so if it is short the message is regarded as not for me.